hal/mbed_ticker_api.c

```diff
@@ -13,61 +13,178 @@
-    data->interface->clear_interrupt();
+void ticker_irq_handler(const ticker_data_t *const ticker) {
+    ticker->interface->clear_interrupt();
 
     /* Go through all the pending TimerEvents */
     while (1) {
-        if (data->queue->head == NULL) {
-            // There are no more TimerEvents left, so disable matches.
-            data->interface->disable_interrupt();
-            return;
+        if (ticker->queue->head == NULL) {
+            break;
         }
 
-        if ((int)(data->queue->head->timestamp - data->interface->read()) <= 0) {
+        // update the current timestamp used by the queue 
+        update_current_timestamp(ticker);
+
+        if (ticker->queue->head->timestamp <= ticker->queue->timestamp) { 
             // This event was in the past:
             //      point to the following one and execute its handler
-            ticker_event_t *p = data->queue->head;
-            data->queue->head = data->queue->head->next;
-            if (data->queue->event_handler != NULL) {
-                (*data->queue->event_handler)(p->id); // NOTE: the handler can set new events
+            ticker_event_t *p = ticker->queue->head;
+            ticker->queue->head = ticker->queue->head->next;
+            if (ticker->queue->event_handler != NULL) {
+                (*ticker->queue->event_handler)(p->id); // NOTE: the handler can set new events
             }
             /* Note: We continue back to examining the head because calling the
              * event handler may have altered the chain of pending events. */
         } else {
-            // This event and the following ones in the list are in the future:
-            //      set it as next interrupt and return
-            data->interface->set_interrupt(data->queue->head->timestamp);
-            return;
-        }
+            break;
+        } 
     }
+
+    update_interrupt(ticker);
 }
 
-void ticker_insert_event(const ticker_data_t *const data, ticker_event_t *obj, timestamp_t timestamp, uint32_t id) {
+void ticker_insert_event(const ticker_data_t *const ticker, ticker_event_t *obj, timestamp_t timestamp, uint32_t id) {
     /* disable interrupts for the duration of the function */
     core_util_critical_section_enter();
 
+    // update the current timestamp
+    update_current_timestamp(ticker);
+    us_timestamp_t absolute_timestamp = convert_relative_timestamp(
+        ticker->queue->timestamp, 
+        timestamp
+    );
+    core_util_critical_section_exit();
+
+    // defer to ticker_insert_event_us
+    ticker_insert_event_us(
+        ticker, 
+        obj, absolute_timestamp, id
+    );
+}
+
+void ticker_insert_event_us(const ticker_data_t *const ticker, ticker_event_t *obj, us_timestamp_t timestamp, uint32_t id) { 
+    /* disable interrupts for the duration of the function */
+    core_util_critical_section_enter();
+
+    // update the current timestamp
+    update_current_timestamp(ticker);
+
+    // filter out timestamp in the past 
```

Comment:
I'm not sure of the desirable behavior here: the event can either be dropped and forget or added to the queue and the queue scheduled immediately. 

@0xc0170 @geky Thoughts ?